File: cups.hs
import Data.Maybe
import System.Environment
import qualified Data.Map as Map
import qualified Data.Sequence as Seq

data Cups = Cups {-# UNPACK #-} !Int
                 {-# UNPACK #-} !Int
    deriving (Eq, Ord)

instance Show Cups where
    show (Cups x y) = "(" ++ show x ++ ", " ++ show y ++ ")"

emptyX :: Int -> Int -> Cups -> Cups
emptyX _ _ (Cups _ y) = Cups 0 y

emptyY :: Int -> Int -> Cups -> Cups
emptyY _ _ (Cups x _) = Cups x 0

fillX :: Int -> Int -> Cups -> Cups
fillX volumeX _ (Cups _ y) = Cups volumeX y

fillY :: Int -> Int -> Cups -> Cups
fillY _ volumeY (Cups x _) = Cups x volumeY

pourXY :: Int -> Int -> Cups -> Cups
pourXY _ volumeY (Cups x y) = Cups (x+y-y') y' where y' = min (x+y) volumeY

pourYX :: Int -> Int -> Cups -> Cups
pourYX volumeX _ (Cups x y) = Cups x' (x+y-x') where x' = min (x+y) volumeX

expand :: Ord a => [a -> a] -> (a -> Bool) -> a -> (Maybe a, Map.Map a a)
expand fneighbors fulfillTarget start =
    expand' (Seq.singleton start, Map.singleton start start)
  where
    expand' (queue, parents) =
        case Seq.viewl queue of
            Seq.EmptyL -> (Nothing, parents)
            x Seq.:< xs -> if fulfillTarget x
                then (Just x, parents)
                else expand' (queue', parents')
              where
                (queue', parents') = foldr expand'' (xs, parents) neighbors
                expand'' = (\ v t@(q, p) -> if Map.member v p
                    then t
                    else (q Seq.|> v, Map.insert v x p))
                neighbors = map ($ x) fneighbors

findPath :: Ord a => [a -> a] -> (a -> Bool) -> a -> Maybe [a]
findPath fneighbors fulfillTarget start =
    maybe Nothing (Just . reverse . findPath') maybeDest
  where
    (maybeDest, parents) = expand fneighbors fulfillTarget start
    findPath' x = if x == start
        then [start]
        else x : (findPath' $ fromJust $ Map.lookup x parents)

findPathCups :: Int -> Int -> Int -> Maybe [Cups]
findPathCups volumeX volumeY targetVolume =
    findPath fneighbors fulfillTarget (Cups 0 0)
  where
    fneighbors = map (\ f -> f volumeX volumeY)
        [emptyX, emptyY, fillX, fillY, pourXY, pourYX]
    fulfillTarget (Cups x y) = any (== targetVolume) [x, y]

main :: IO ()
main = do
    args <- fmap (map read) getArgs
    let maybePath = findPathCups
            (case drop 0 args of x:_ -> x; [] -> 4)
            (case drop 1 args of x:_ -> x; [] -> 9)
            (case drop 2 args of x:_ -> x; [] -> 2)
    case maybePath of
        Nothing -> putStrLn "No possible solution!"
        Just path -> do
            let nstep = length path - 1
            putStrLn $ "Found an answer with " ++ show nstep ++ " steps!"
            mapM_ print path

File: cups.py
import sys

MAX_VOL_X = int(sys.argv[1]) if len(sys.argv) >= 2 else 9
MAX_VOL_Y = int(sys.argv[2]) if len(sys.argv) >= 3 else 4
TARGET_VOLUME = int(sys.argv[3]) if len(sys.argv) >= 4 else 2

class CupState:
	volume_x = 0
	volume_y = 0
	parent = False

	def __init__(self , vol_x , vol_y , parent_state):
		self.volume_x = vol_x
		self.volume_y = vol_y
		self.parent = parent_state

	def empty_x(self):
		return CupState(0 , self.volume_y , self)

	def empty_y(self):
		return CupState(self.volume_x , 0 , self)

	def fill_x(self):
		return CupState(MAX_VOL_X , self.volume_y , self)

	def fill_y(self):
		return CupState(self.volume_x , MAX_VOL_Y , self)

	def pour_x_to_y(self):
		y = min(MAX_VOL_Y , self.volume_y + self.volume_x)
		x = self.volume_x - (y-self.volume_y)
		return CupState(x, y, self)

	def pour_y_to_x(self):
		x = min(MAX_VOL_X , self.volume_y + self.volume_x)
		y = self.volume_y - (x-self.volume_x)
		return CupState(x, y, self)

	def isTargetVolume(self):
		return self.volume_y == TARGET_VOLUME or self.volume_x == TARGET_VOLUME

bfs_queue = [];
initial_state = CupState(0 , 0 , False)
bfs_queue.append(initial_state)
visited_states = [[0 for x in range(MAX_VOL_X + 1)] for y in range(MAX_VOL_Y + 1)]

while len(bfs_queue) > 0:
	current_state = bfs_queue.pop(0)
	if visited_states[current_state.volume_y][current_state.volume_x]:
		continue

	if current_state.isTargetVolume():
		num_x = []
		num_y = []
		while True:
			num_x.append(current_state.volume_x)
			num_y.append(current_state.volume_y)
			current_state = current_state.parent
			if current_state == False:
				break

		print("Found a solution with %s steps!" % (len(num_x) - 1))
		while len(num_x):
			print("(%s , %s)" % (num_x.pop() , num_y.pop()))
		exit()

	visited_states[current_state.volume_y][current_state.volume_x] = True
	for mutation in range(0,6):

		if mutation == 0: next_state = current_state.empty_x()
		elif mutation == 1: next_state = current_state.empty_y()
		elif mutation == 2: next_state = current_state.pour_x_to_y()
		elif mutation == 3: next_state = current_state.pour_y_to_x()
		elif mutation == 4: next_state = current_state.fill_x()
		elif mutation == 5: next_state = current_state.fill_y()
		#print next_state
		#print len(visited_states), len(visited_states[0]), next_state.volume_y, next_state.volume_x
		if visited_states[next_state.volume_y][next_state.volume_x] == False:
			bfs_queue.append(next_state)

print("No possible solutions!")
